packages/commonwealth/client/scripts/state/api/comments/editComment.ts

```diff
@@ -0,0 +1,89 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import axios from 'axios';
+import Comment from 'models/Comment';
+import app from 'state';
+import { ApiEndpoints } from 'state/api/config';
+import useFetchCommentsQuery from './fetchComments';
+
+interface EditCommentProps {
+  address: string;
+  chainId: string
+  parentCommentId: number | null,
+  threadId: number;
+  commentId: number;
+  updatedBody: string;
+  attachments?: string[];
+}
+
+const editComment = async ({
+  address,
+  chainId,
+  parentCommentId,
+  threadId,
+  commentId,
+  updatedBody,
+  attachments,
+}: EditCommentProps) => {
+  const {
+    session = null,
+    action = null,
+    hash = null,
+  } = await app.sessions.signComment({
+    thread_id: threadId,
+    body: updatedBody,
+    parent_comment_id: parentCommentId,
+  })
+
+  const response = await axios.patch(`${app.serverUrl()}/comments/${commentId}`, {
+    address: address,
+    author_chain: chainId,
+    id: commentId,
+    chain: chainId,
+    body: encodeURIComponent(updatedBody),
+    'attachments[]': attachments,
+    jwt: app.user.jwt,
+    canvas_action: action,
+    canvas_session: session,
+    canvas_hash: hash,
+  })
+
+  return new Comment(response.data.result)
+};
+
+interface UseEditCommentMutationProps {
+  chainId: string
+  threadId: number;
+}
+
+const useEditCommentMutation = ({ chainId, threadId }: UseEditCommentMutationProps) => {
+  const queryClient = useQueryClient();
+  const { data: comments } = useFetchCommentsQuery({
+    chainId,
+    threadId,
+  })
+
+  return useMutation({
+    mutationFn: editComment,
+    onSuccess: async (updatedComment) => {
+      // find the existing comment index
+      const foundCommentIndex = comments.findIndex(x => x.id === updatedComment.id)
+
+      // update fetch comments query state with updated comment
+      if (foundCommentIndex > -1) {
+        const key = [ApiEndpoints.FETCH_COMMENTS, chainId, threadId]
+        queryClient.cancelQueries({ queryKey: key });
+        queryClient.setQueryData([...key],
+          () => {
+            const updatedComments = [...(comments || [])]
+            updatedComments[foundCommentIndex] = updatedComment
+            return [...updatedComments]
```

Comment:
we keep repeating this pattern of adding, editing, removing items from the list. Would be nice to have some utils that will wrap the logic into single line. 

Also wondering if we really need to return non-mutable data? 
What if we do this as I showed above

```   
 return comments.map(comment => {
      if (comment.id === updatedComment.id) {
        return updatedComment
      }
      
      return comment
    })  
```

Reply:
I believe if we mutate the data it will mess up the react query cache and since the cache is internal, in case of a bug it will be hard to track if mutated.

I went little overboard with the `...`, the last `return [...updatedComments]` can be removed